Support custom resttemplate

natrem opened the pull request:
as discussed in #1327 if I want to use a specific `RestTemplate` extension (e.g. a `OAuth2RestTempalte`) then I have to duplicate a lot of code from spring-cloud-sleuth.
This is due to many classes and Configuration classes being package private.

In this PR I add 2 things:
- the default `ZipkinRestTemplateWrapper` is now exposed as a bean with specific name (similar as Sender bean)
- since it is a bean it is subject to instrumentation by `TraceWebClientAutoConfiguration`; I added a utility class to filter the brave interceptors. This class is used in `ZipkinRestTemplateWrapper` to avoid tracing sending the spans and remains public for users wanting to replace the rest template.

I you feel like making this RestTemplate a bean is too risky then we should make several classes public to allow reuse of the default features in custom Senders.

Notes on the tests:
- I included an example of invalid extension to show the risk.
- I included an example of valid extension. this one might be dropped if you consider that the invalid test and the simple test showing the interceptors removal is enough
- these two tests are a bit slow (5s each)
- I did not manage to have `TraceWebClientAutoConfiguration` work properly in the unit test so I reproduced it's behavior in the test.

Notes on the documentation:
- I tried to update it but did not check the result at the time of this writing

devinsba commented on spring-cloud-sleuth-zipkin/src/main/java/org/springframework/cloud/sleuth/zipkin2/sender/ZipkinRestTemplateSenderConfiguration.java:
```diff
@@ -34,30 +34,46 @@
+	 * Zipkin rest template bean name. Name of the bean matters for supporting multiple tracing
+	 * systems.
+	 */
+	public static final String REST_TEMPLATE_BEAN_NAME = "zipkinRestTemplate";
```
I think your intent is that this is public so you should move it to `ZipkinAutoConfiguration` with the other constants like this

natrem replied:
done

devinsba commented on docs/src/main/asciidoc/intro.adoc:
```diff
@@ -429,6 +429,18 @@ To do this you can use respectively `ZipkinAutoConfiguration.REPORTER_BEAN_NAME`
+The default web sender uses a specific `RestTemplate` that you can customize by providing a bean of type
+`ZipkinRestTemplateCustomizer`. If you wish to use a more specific RestTemplate (e.g. a OAUth2RestTemplate) 
+then you need to give your bean the specific name 
+`ZipkinRestTemplateSenderConfiguration.REST_TEMPLATE_BEAN_NAME`. 
```
Update this when you move the property to the other class

natrem replied:
done

natrem commented:
While fixing the forgotten integration test (commit b4743959052254260723b057a67e5e02cf5ffb7d) I updated `AssertingRestTemplate` and moved the utility class to spring-cloud-sleuth-core module.

We could have an annotation `SleuthWebClientExcluded` to ask sleuth no to add the interceptors instead of rejecting them with the utility class but I don't think it simplifies the use case.

natrem commented:
Is there something I can help with to help you decide on the value of this PR for the project? Or did I miss something?

marcingrzejszczak commented:
Hi @natrem . Thank you for this PR. I didn't have time to look into it yet. I have some mixed feelings about `SleuthWebClientInterceptorRemover` and `SleuthWebClientExcluded`. I haven't however spent any time on suggesting a better approach.

natrem commented:
sorry if I seemed pushy. I understand your concern - that I share.
The main other possibility I considered was to simply make some classes public to ease creation of custom Sender `ZipkinAutoConfiguration.SENDER_BEAN_NAME`
Another possibility would be to create a class similar to ZipkinRestTemplateWrapper but using an `OAuth2RestTemplate` if there is a bean for the Oauth2 configuration (or is properties are filled)
Would you rather explore this possibility?

marcingrzejszczak commented:
> sorry if I seemed pushy. I understand your concern - that I share.

Don't worry, you weren't.

> Another possibility would be to create a class similar to ZipkinRestTemplateWrapper but using an OAuth2RestTemplate if there is a bean for the Oauth2 configuration (or is properties are filled)

That sounds much better IMO.

>  Would you rather explore this possibility?

Absolutely. Be my guest.

natrem commented:
Adaptation proposed in #1362

spencergibb commented:
Should this be closed then?

natrem commented:
I'd say that it depends on the discussion of new proposal, but most probably yes. We'll reopen if needed.
